### PR TITLE
Respect object property hp_max field for players

### PR DESCRIPTION
**Pull request by SmallJoker**
This allows modders to configure the maximal HP per player.

Apart from #2246 there's surely another request for this around.
```Lua
player:set_properties({
	hp_max = 30
})
```
With too high values, the health bar will go off-screen when fully "recharged".

### CODE REVIEW

**Comment by lisacvuk** [comment]
awesome, I really need this for my mod. Thanks! :)

**Comment by nerzhul** [comment]
it seems okay except we have a sync problem with the hardcoded define on client, did you tested it with 15 hp over 100 mx hp to see if you have 15% hp bar on HUD ?

**Comment by SmallJoker** [comment]
@nerzhul for once, this is a server sided only feature. Each heart in the HUD represents two HP. If the maximal HP limit is raised, the size of this HUD element will increase. To not overlap with breath, modders can change this display type to a bar or similar.

**Comment by nerzhul** [comment]
don't forget HP Hud is embbeded in MT core if i'm exact, it's not exactly what you think :)

**Comment by SmallJoker** [comment]
grafik
Tweaked hudbars to deal with the high HP values. The red bar offset seems to be a bug in this mod.

**Comment by nerzhul** [comment]
can you fix the default representation to prevent a such overflow ?

File: builtin/game/statbars.lua [hunk 1]
```diff
@@ -37,39 +37,44 @@ local function initialize_builtin_statbars(player)
...
 	end
+	local hud = hud_ids[name]
 
 	if player:hud_get_flags().healthbar and enable_damage then
- 		if hud_ids[name].id_healthbar == nil then
-			health_bar_definition.number = player:get_hp()
-			hud_ids[name].id_healthbar  = player:hud_add(health_bar_definition)
+ 		if hud.id_healthbar == nil then
+ 			local hp = player:get_hp()
+ 			local max_display_hp = math.max(20, math.max(
```
Comment (nerzhul) [inline builtin/game/statbars.lua]: maybe having a constant somewhere for the 20 should be nice

File: builtin/game/statbars.lua [hunk 2]
```diff
@@ -37,39 +37,44 @@ local function initialize_builtin_statbars(player)
...
-			player:hud_remove(hud_ids[name].id_healthbar)
-			hud_ids[name].id_healthbar = nil
+		if hud.id_healthbar ~= nil then
+			player:hud_remove(hud.id_healthbar)
+			hud.id_healthbar = nil
 		end
 	end
 
-	if (player:get_breath() < 11) then
+	if player:get_breath() < 11 then
```
Comment (nerzhul) [inline builtin/game/statbars.lua]: same for this constant

File: builtin/game/constants.lua [hunk 3]
```diff
@@ -21,6 +21,10 @@ core.EMERGE_GENERATED   = 4
 -- constants.h
 -- Size of mapblocks in nodes
 core.MAP_BLOCKSIZE = 16
+-- Default maximal HP of a player
+core.PLAYER_MAX_HP = 20
```
Comment (nerzhul) [inline builtin/game/constants.lua]: DEFAULT_PLAYER_MAX_HP ? same for breath ?

**Comment by nerzhul** [comment]
Okay for me

**Pull request merged**